### PR TITLE
[Model][PP] Relay DeepSeek-V4.1 cache and index state across stages

### PR DESCRIPTION
**Pull request by 0z5a**
## Purpose

Add the default-off `--additional-config '{"deepseek_v41_pp_sharing":true}'` eager V1 path. Register cache replicas and relay active KV/index-K blocks, top-k indices and candidate blocks across pipeline stages, including intermediate relay stages. Preserve cache layout and model-owned payload axes instead of treating every intermediate tensor as token-indexed. Bound each outgoing snapshot with `deepseek_v41_pp_share_max_bytes` (default 512 MiB).

This includes the dependency discovery and local-cache allocation prerequisite from #56221. #56222 supplies the independent PP+SP stage-boundary changes. The implementation rejects microbatching, context parallelism, speculative decoding, KV transfer and external-launcher execution. The duplicate check found no other open V4.1 PR implementing this complete dependency-aware relay; the validation-only #56221 does not transfer state.

Based on #56214 (`dsv41-feat`, `c9d909e802a39292f54101bff8a36096761ea605`). The PR targets that feature branch so the model implementation is not repeated in this diff.

Rebased on 2026-09-11 after the feature branch was rewritten. Changed-file Python parsing and all applicable pre-commit hooks passed. The pipeline suites passed (20 CPU tests, including the real three-process Gloo relay). GPU/full-model measurements below belong to the prior revision based on `e47aa780bccf59f59dfa2cbb18e17a10b4fe69ba`; they were not rerun on this rebased head.

AI assistance: OpenAI Codex assisted with implementation, review, test execution and preparation of this PR. This draft does not claim that a human has completed a line-by-line review.

## Test Plan

Fresh split: **20 CPU tests passed**, including the actual three-process Gloo/two-hop relay, packed/strided cache snapshots, route deduplication, empty snapshots, budget failures and layout checks. The dependency-light loader runs the actual modules and tensor/collective operations without CUDA model-registry imports:

```text
.venv/bin/python run_cpu_checks.py tests/models/test_deepseek_v41_pipeline.py \
  tests/models/test_deepseek_v41_pipeline_transfer.py \
  tests/distributed/test_deepseek_v41_pipeline_transfer.py
```

All applicable pre-commit hooks and changed-file Python parsing passed.

## Test Result

Recorded full-model runs used `deepseek-ai/DeepSeek-V4.1-Flash`, revision `df42c109f1defefcbfcedbe7d905718a12266e40`, on 4×SXM H100 80 GiB. Each positive matrix case checked six short questions, cold/cached long input with chunked prefill, and four unequal-length concurrent requests. “Text equal” means short-answer text equality to the TP4 reference, not logits/token-probability parity or a standard accuracy benchmark.

**Evidence scope:** GPU results below come from the earlier combined integration based on #56214, including companion changes. The isolated PR has not had a new full-model GPU run. Fresh split checks are listed separately; passing the integrated run does not establish isolated-branch equivalence.

| Integrated configuration | QA answers correct | QA text equal to TP4 | Long / chunked + prefix | Concurrent requests | Strict result |
| --- | --- | --- | --- | --- | --- |
| PP4×TP1 sharing eager, 10/10/10/10 | 6/6 | 6/6 | 2/2 correct; cache reuse checked | 4/4 correct | PASS |
| PP2×TP2 + SP (#56222) + sharing eager, 19/21 | 6/6 | 6/6 | 2/2 correct; cache reuse checked | 4/4 correct | PASS |
| PP2×TP2 + SP + sharing + overlap (#56220), 19/21 | 6/6 | 5/6 | 2/2 correct; cache reuse checked | 4/4 correct | Text mismatch; strict check not passed |

The overlap combination has a sequence-answer wording difference and is not counted as strict equivalence.

**Short-request serving — one run per configuration**

| Integrated configuration | Layer partition | Input → output | Successful requests / concurrency | Output tokens/s | Median TTFT ms | Median TPOT ms | Median request latency ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| PP4×TP1 sharing | 10/10/10/10 | 512 → 128 | 16/16; 4 | 5.920 | 20934.75 | 513.89 | 86517.38 |
| PP2×TP2+SP sharing | 19/21 | 512 → 128 | 16/16; 4 | 11.458 | 8235.85 | 294.05 | 44472.22 |

Eager, CPU weight offload 8 GiB, Engram CPU offload, prefix cache off, four warmup requests, fixed input seed 42, and 128 output tokens. These are absolute supported-configuration results: sharing-off rejects both partitions, so no same-partition A/B speedup can be inferred. One run cannot quantify run-to-run variance. No memory-saving figure is established for sharing or PP+SP.

| Fresh split validation | Result | Coverage |
| --- | --- | --- |
| Standalone CPU tests | 20/20 PASS | Dependency routes, packed/strided snapshots, deduplication, empty payloads, budgets and layouts |
| Real distributed transfer, included in the 20 tests | PASS | 3 Gloo processes, 2-hop relay; actual tensor/collective operations |
| Applicable pre-commit hooks and changed Python parsing | PASS; 13 changed files | Isolated PR source |

---
<details>
<summary>PR description checklist</summary>

- [x] Purpose and related work described.
- [x] Test plan and actual results stated.
- [x] Model evaluation limitations stated.
- [x] AI assistance disclosed.

</details>

<details>
<summary>Standalone CPU validation loader</summary>

Saved as `run_cpu_checks.py`; set `DSV41_SPLIT_REPO` to the checked-out branch and pass the test paths listed above. This avoids CUDA model-registry imports on macOS while executing the actual leaf-module implementation, tests and tensor/collective operations. It is not a CUDA or full-model test.

```python
"""Run unchanged CPU test functions against split source without CUDA imports.

Preload only dependency-light leaf modules, as in the integration validation
driver. CUDA-only tests keep their own skip conditions. No tensor operations or
collectives are mocked by this loader.
"""
import importlib.util
import os
from pathlib import Path
import sys

import pytest

ROOT = Path(os.environ["DSV41_SPLIT_REPO"])
for name in (
    "vllm.models.deepseek_v4_1.common.pipeline",
    "vllm.models.deepseek_v4_1.common.pipeline_transfer",
    "vllm.v1.worker.ubatch_inputs",
):
    path = ROOT / (name.replace(".", "/") + ".py")
    if not path.is_file():
        continue
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    sys.modules[name] = module
    spec.loader.exec_module(module)

if __name__ == "__main__":
    raise SystemExit(pytest.main(["-q", "--noconftest", "-o", "addopts=", *sys.argv[1:]]))
```

</details>

### CODE REVIEW

**Comment by 0z5a** [comment]
Validation follow-up for `dfe0c6f292703b22b933a72d3db945d1f6785b0b`: the 20 current-head CPU tests, including a real two-hop Gloo relay, are useful. The remaining gap is an isolated current-head full-model GPU run; the recent H200 session ended before this case started.

The highest-value contribution would exercise eager V1 PP4×TP1 with the 10/10/10/10 partition and prove that real requests transfer nonempty KV, index-K, top-k and candidate payloads through every required hop. Check the received/restored bytes, long/chunked requests, prefix-cache reuse, changing request layouts and the configured byte-budget guards. Benchmark supported configurations and quantify transfer/allocation costs. Since sharing-off rejects this partition, a same-partition off/on throughput ratio is not a valid comparison; this feature can be judged on correct support and measured cost without inventing a speedup.

Contributions from anyone with access to sufficient GPU capacity are very welcome: reproducible GPU E2E benchmarks, targeted correctness checks, and focused fixes. Please include source/model revisions, GPU topology, offload/KV settings, exact commands, and raw results; repeat timing runs when making performance claims. I am happy to review and cherry-pick fixes while preserving the original authorship and clear attribution.

**Comment by 0z5a** [comment]
Resubmitted as #56439 against `main`, since `dsv41-feat` was merged into `main` in #56214.
